File: save-me-chunker/src/ChunkProtocol.js
import { normalizeNewlines, stripCodeFences } from './utils.js';

// Message formats:
// [FILE START]\n
// id=<uuid>\n
// path=<dst/relative.ext>\n
// total_parts=<N>\n
// encoding=base64\n
// compression=<none|gzip>\n
// sha256=<hash>
//
// [FILE CHUNK i/N id=<uuid>]\n
// <base64 data>
//
// [FILE END id=<uuid>]
// [FILE CANCEL id=<uuid>]

const START_HEADER = '[FILE START]';
const END_PREFIX = '[FILE END';
const CANCEL_PREFIX = '[FILE CANCEL';
const CHUNK_PREFIX = '[FILE CHUNK';

export function parseProtocolMessage(rawText) {
  if (!rawText || typeof rawText !== 'string') return null;
  let text = normalizeNewlines(stripCodeFences(rawText)).trim();

  if (text.startsWith(START_HEADER)) {
    const body = text.slice(START_HEADER.length).trim();
    const meta = parseKeyValueBody(body);
    if (!meta.id) return { type: 'error', error: 'missing id in START' };
    const total = parseInt(meta.total_parts, 10);
    if (!Number.isFinite(total) || total <= 0) return { type: 'error', error: 'invalid total_parts' };
    const encoding = (meta.encoding || 'base64').toLowerCase();
    const compression = (meta.compression || 'none').toLowerCase();
    return {
      type: 'start',
      id: meta.id,
      meta: {
        path: meta.path,
        total_parts: total,
        encoding,
        compression,
        sha256: meta.sha256 || null,
      },
    };
  }

  if (text.startsWith(CHUNK_PREFIX)) {
    // [FILE CHUNK i/N id=<uuid>]\n<base64>
    const headerMatch = text.match(/^\[FILE CHUNK\s+(\d+)\s*\/\s*(\d+)\s+id=([^\]]+)\]\s*\n([\s\S]*)$/);
    if (!headerMatch) return { type: 'error', error: 'invalid CHUNK header' };
    const index = parseInt(headerMatch[1], 10);
    const total = parseInt(headerMatch[2], 10);
    const id = headerMatch[3].trim();
    const data = headerMatch[4].trim();
    return { type: 'chunk', id, index, total, dataB64: data };
  }

  if (text.startsWith(END_PREFIX)) {
    const m = text.match(/^\[FILE END\s+id=([^\]]+)\]\s*$/);
    if (!m) return { type: 'error', error: 'invalid END header' };
    return { type: 'end', id: m[1].trim() };
  }

  if (text.startsWith(CANCEL_PREFIX)) {
    const m = text.match(/^\[FILE CANCEL\s+id=([^\]]+)\]\s*$/);
    if (!m) return { type: 'error', error: 'invalid CANCEL header' };
    return { type: 'cancel', id: m[1].trim() };
  }

  return null;
}

function parseKeyValueBody(body) {
  const meta = {};
  const lines = body.split(/\n+/).map((s) => s.trim()).filter(Boolean);
  for (const line of lines) {
    const eqIdx = line.indexOf('=');
    if (eqIdx === -1) continue;
    const key = line.slice(0, eqIdx).trim();
    const value = line.slice(eqIdx + 1).trim();
    meta[key] = value;
  }
  return meta;
}

export function formatStart({ id, path, total_parts, encoding = 'base64', compression = 'none', sha256 = null }) {
  let out = `${START_HEADER}\n`;
  out += `id=${id}\n`;
  out += `path=${path}\n`;
  out += `total_parts=${total_parts}\n`;
  out += `encoding=${encoding}\n`;
  out += `compression=${compression}\n`;
  if (sha256) out += `sha256=${sha256}\n`;
  return out.trim();
}

export function formatChunk({ id, index, total, dataB64 }) {
  return `[FILE CHUNK ${index}/${total} id=${id}]\n${dataB64}`;
}

export function formatEnd({ id }) {
  return `[FILE END id=${id}]`;
}

export function formatCancel({ id }) {
  return `[FILE CANCEL id=${id}]`;
}

File: src/chunk_protocol.py
import re
from typing import Optional, Dict, Any

START_HEADER = '[FILE START]'

_CHUNK_RE = re.compile(r"^\[FILE CHUNK\s+(\d+)\s*\/\s*(\d+)\s+id=([^\]]+)\]\s*\n([\s\S]*)$", re.MULTILINE)
_END_RE = re.compile(r"^\[FILE END\s+id=([^\]]+)\]\s*$", re.MULTILINE)
_CANCEL_RE = re.compile(r"^\[FILE CANCEL\s+id=([^\]]+)\]\s*$", re.MULTILINE)


def _strip_code_fences(text: str) -> str:
    text = text.strip()
    if text.startswith('```') and text.endswith('```'):
        inner = text.split('\n', 1)
        if len(inner) == 2:
            body = inner[1]
            if body.endswith('```'):
                body = body[:-3]
            return body.strip()
    return text


def parse_protocol_message(raw_text: str) -> Optional[Dict[str, Any]]:
    if not isinstance(raw_text, str) or not raw_text:
        return None
    text = _strip_code_fences(raw_text).strip()

    if text.startswith(START_HEADER):
        body = text[len(START_HEADER):].strip()
        meta: Dict[str, str] = {}
        for line in body.split('\n'):
            line = line.strip()
            if not line or '=' not in line:
                continue
            key, value = line.split('=', 1)
            meta[key.strip()] = value.strip()
        if 'id' not in meta:
            return {"type": "error", "error": "missing id in START"}
        try:
            total = int(meta.get('total_parts', '0'))
        except ValueError:
            total = 0
        if total <= 0:
            return {"type": "error", "error": "invalid total_parts"}
        encoding = (meta.get('encoding') or 'base64').lower()
        compression = (meta.get('compression') or 'none').lower()
        return {
            'type': 'start',
            'id': meta['id'],
            'meta': {
                'path': meta.get('path'),
                'total_parts': total,
                'encoding': encoding,
                'compression': compression,
                'sha256': meta.get('sha256') or None,
            }
        }

    m = _CHUNK_RE.match(text)
    if m:
        index = int(m.group(1))
        total = int(m.group(2))
        id_ = m.group(3).strip()
        data = m.group(4).strip()
        return { 'type': 'chunk', 'id': id_, 'index': index, 'total': total, 'dataB64': data }

    m = _END_RE.match(text)
    if m:
        return { 'type': 'end', 'id': m.group(1).strip() }

    m = _CANCEL_RE.match(text)
    if m:
        return { 'type': 'cancel', 'id': m.group(1).strip() }

    return None